.circleci/config.yml

```diff
@@ -1,106 +1,92 @@
 ---
 version: 2.1
-orbs:
-  win: circleci/windows@2.4.0
+
+parameters:
+  go-version:
+    type: string
+    default: "1.14.2"
+  workspace-dir:
+    type: string
+    default: "/home/circleci"
+
+commands:
+  setup:
+    description: "install go, checkout and restore cache"
+    steps:
+      - run:
+          name: "install go"
+          command: |
+            curl --create-dirs -o $GOPATH/go.tar.gz https://dl.google.com/go/go${GOVERSION}.linux-amd64.tar.gz
+            tar --strip-components=1 -C $GOPATH -xzf $GOPATH/go.tar.gz
+            rm -rf $GOPATH/go.tar.gz
+      - checkout
+      - restore_cache:
+          keys:
+            - 'v2-pkg-cache-{{ checksum "go.sum" }}-{{ .Environment.GOVERSION }}'
+
+executors:
+  linux:
+    machine:
+      image: ubuntu-1604:201903-01
+      docker_layer_caching: true
+    working_directory: << pipeline.parameters.workspace-dir >>/project
+    environment:
+      GOPATH: << pipeline.parameters.workspace-dir >>/go/<< pipeline.parameters.go-version >>
+      PATH: << pipeline.parameters.workspace-dir >>/go/<< pipeline.parameters.go-version >>/bin:/usr/local/bin:/usr/bin:/bin
+      GOVERSION: << pipeline.parameters.go-version >>
 
 workflows:
   version: 2
-  test_os:
+  main:
     jobs:
-      - go-install
-      - tidy-lint:
+      - prepare-tidy
+      - lint:
           requires:
-            - go-install
-      - test-linux:
+            - prepare-tidy
+      - build-test:
           requires:
-            - go-install
-      - test-windows
+            - prepare-tidy
+
 jobs:
-  go-install:
-    machine:
-      image: ubuntu-1604:201903-01
+  prepare-tidy:
+    executor: linux
     steps:
-      - restore_cache:
-          keys:
-            - v1-pkg-cache
+      - setup
       - run:
-          name: "install go 1.14"
-          command: |
-            ls -A ~/go && echo already downloaded. exiting. && exit 0
-            curl -o ~/go.tar.gz https://dl.google.com/go/go1.14.2.linux-amd64.tar.gz
-            tar -C ~/ -xzvf ~/go.tar.gz
-
+          name: "make tidy"
+          command: make tidy && git diff --exit-code
+      - run:
+          name: "make mod-download"
+          command: make mod-download
       - save_cache:
-          key: v1-pkg-cache
+          key: 'v2-pkg-cache-{{ checksum "go.sum" }}-{{ .Environment.GOVERSION }}'
           paths:
-            - ~/go
+            - << pipeline.parameters.workspace-dir >>/go/<< pipeline.parameters.go-version >>/pkg/mod
```

Comment:
I was hoping to avoid having to download go installation every time. What's faster and more reliable-- restoring cache or downloading from dl.google.com?

Reply:
There are workspaces and caches.

Caches survive builds, and we _could_ cache go _as long as_ we embed the go version inside the cache key. We were not doing that before, so even if we upgraded the go version, we would still be left with the old one.

Caches are fast-ish. In practice, it takes 10 sec to download and unpack go. Adding the go distribution to the cache will bloat it up by ~100M, which will also add up to the cache download... All in all, we're looking at single digits seconds gains, not worth it.

OTOH, workspaces are storage area for a single workflow, shared across jobs. They are DAMN SLOW; stay away.